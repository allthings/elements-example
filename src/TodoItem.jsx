import React from 'react'
import PropTypes from 'prop-types'
import Icon from '@allthings/elements/atoms/Icon'
import Text from '@allthings/elements/atoms/Text'
import Checkmark from '@allthings/elements/molecules/Checkmark'
import ListItem from '@allthings/elements/molecules/List/ListItem'
import TextInput from '@allthings/elements/molecules/TextInput'
import { css } from 'glamor'

class TodoItem extends React.Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    children: PropTypes.string.isRequired,
    done: PropTypes.bool.isRequired
  }

  state = {
    touchStart: undefined,
    icons: 0,
    iconTransition: '',
    slideLeft: -70,
    checkmark: this.props.checkmarkDisplay || 'flex',
  }

  handleRemove = () => {
    this.closeIcons('')
    this.props.handleRemove(this.props.id) 
  }

  handleEdit = event => {
    this.closeIcons('')
    this.props.onDoubleClick(this.props.id)
  }

  handleEditComplete = event => {
    this.props.doneEditting(event, this.props.id)
    this.closeIcons('')
  }

  handleCheckClick = () => this.props.checkmarkClicked(this.props.id)

  handleTouchStart = event => {
    const touchX = event.changedTouches[0].clientX
    this.setState({ touchStart: touchX })
  }

  openIcons = () => {
    this.setState({
      checkmark: 'none',
      slideLeft: 0,
      icons: 150,
      iconTransition: ''
    })
  }

  closeIcons = transition => {
    this.setState({
      checkmark: 'flex',
      slideLeft: -70,
      icons: 0,
      iconTransition: transition || '',
      doubleClicked: false,
    })
  }

  handleTouchMove = event => {
    event.preventDefault()
    const { slideLeft, touchStart } = this.state
    const { onSlideLeft, id } = this.props
    const touchObj = event.changedTouches[0].clientX
    const dist = parseInt(touchObj,) - touchStart
    if (slideLeft < 0 && dist < -10) {
      this.openIcons()
      onSlideLeft(id, 'open')
    } else if (dist > 5) {
      this.closeIcons('200ms')
      onSlideLeft(id, 'close')
    }
  }

  render() {
    const { id, children, done, doubleClicked, iconOpen } = this.props
    const { icons, iconTransition, slideLeft, checkmark } = this.state
    const checkmarkNo = (iconOpen || doubleClicked) ? 'none' : checkmark
    const styles = {
      check: css({
        height: '40px',
        width: '40px',
        display: checkmarkNo + ' !important',
        margin: '3px 0px'
      }),
      inline: css({
        overflow: 'hidden',
        display: 'inline-flex',
        width: icons,
        transitionDelay: iconTransition
      }),
      list: css({
        display: 'inline-flex'
      }),
      listItem: css({
        minHeight: '50px',
        width: '100%'
      }),
      task: css({
        display: 'inline-flex',
        width: '90%',
        padding: '10px 0'
      }),
      text: css({
        margin: '0px 10px',
        width: '100%',
        height: '50.5px',
        display: 'table'
      }),
      removeDiv: css({
        position: 'relative',
        overflow: 'hidden',
        height: '50.5px',
        width: '53.5px',
        padding: '0px 10px 15px 10px'
      }),
      editDiv: css({
        position: 'relative',
        overflow: 'hidden',
        height: '50.5px',
        width: '53.5px',
        padding: '0px 10px 15px 10px'
      }),
      remove: css({
        float: 'right',
        padding: '5px',
        margin: '3px',
        right: slideLeft,
        transition: 'all 300ms ease',
        ':hover': { cursor: 'pointer' }
      }),
      iconBackground: css({
        position: 'absolute',
        padding: '15px',
        float: 'right',
        right: slideLeft,
        transition: 'all 300ms ease',
        height: '35.5px',
        width: '40px'
      }),
      gray: css({
        backgroundColor: 'gray',
        transition: 'all 300ms ease',
      }),
      red: css({
        backgroundColor: 'red',
        transition: 'all 300ms ease',
      }),
      editField: css({
        padding: 0,
        ':focus': { outline: 'none' }
      }),
      staticText: css({
        display: 'table-cell !important',
        verticalAlign: 'middle'
      })
    }

    return (
      <div {...styles.list}>
        <ListItem {...styles.listItem}>
          <div {...styles.task}>
            <Checkmark id={`checkmark${id}`} checked={done} onClick={this.handleCheckClick} {...styles.check} />
            <div {...styles.text} onDoubleClick={this.handleEdit} onTouchStart={this.handleTouchStart} onTouchMove={this.handleTouchMove} >
              {doubleClicked ? <TextInput id={id} name="edit" defaultValue={children} placeholder="Edit todo" onKeyPress={this.handleEditComplete} {...styles.editField} /> : <Text autoBreak={true} id={id} {...styles.staticText}>{children}</Text>}
            </div>
          </div>
          <div {...styles.inline}>
            <div {...styles.editDiv}>
              <div  {...styles.iconBackground} {...styles.gray}>
                <Icon name="edit" {...styles.remove} size="m" color="white" onClick={this.handleEdit} />
              </div>
            </div>
            <div {...styles.removeDiv}>
              <div {...styles.iconBackground} {...styles.red}>
                <Icon name="remove-light-filled" {...styles.remove} size="m" color="white" onClick={this.handleRemove} />
              </div>
            </div>
          </div>
        </ListItem>
      </div>
    )
  }
}

export default TodoItem