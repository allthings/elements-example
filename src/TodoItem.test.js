import React from 'react'
import { shallow, mount, render } from 'enzyme'
import TodoItem from './TodoItem'

describe('TodoItem', () => {
  it('it should match snapshot', () => {
    const wrapper = shallow(<TodoItem id={0} done={false}>Some Text</TodoItem>)
    expect(wrapper).toMatchSnapshot();
  })
})
