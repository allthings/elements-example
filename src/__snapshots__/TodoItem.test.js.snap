// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem it should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoItem
    done={false}
    id={0}
  >
    Some Text
  </TodoItem>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ListItem
        backgroundColor="white"
        data-css-zh5k6q=""
        padded={true}
      >
        <div
          data-css-154b89o=""
        >
          <_class
            checked={false}
            data-css-1mql7k=""
            id="checkmark0"
            onClick={[Function]}
          />
          <div
            data-css-leaz66=""
            onDoubleClick={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
          >
            <Text
              autoBreak={true}
              block={true}
              color="text"
              data-css-93utwa=""
              id={0}
              size="l"
            >
              Some Text
            </Text>
          </div>
        </div>
        <div
          data-css-m0lcdg=""
        >
          <div
            data-css-b7lh05=""
          >
            <div
              data-css-1ajzdj4=""
              data-css-pmoomt=""
            >
              <Icon
                color="white"
                data-css-1lke24q=""
                name="edit"
                onClick={[Function]}
                size="m"
              />
            </div>
          </div>
          <div
            data-css-b7lh05=""
          >
            <div
              data-css-1ajzdj4=""
              data-css-83ugar=""
            >
              <Icon
                color="white"
                data-css-1lke24q=""
                name="remove-light-filled"
                onClick={[Function]}
                size="m"
              />
            </div>
          </div>
        </div>
      </ListItem>,
      "data-css-h0biv2": "",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "backgroundColor": "white",
        "children": Array [
          <div
            data-css-154b89o=""
          >
            <_class
              checked={false}
              data-css-1mql7k=""
              id="checkmark0"
              onClick={[Function]}
            />
            <div
              data-css-leaz66=""
              onDoubleClick={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <Text
                autoBreak={true}
                block={true}
                color="text"
                data-css-93utwa=""
                id={0}
                size="l"
              >
                Some Text
              </Text>
            </div>
          </div>,
          <div
            data-css-m0lcdg=""
          >
            <div
              data-css-b7lh05=""
            >
              <div
                data-css-1ajzdj4=""
                data-css-pmoomt=""
              >
                <Icon
                  color="white"
                  data-css-1lke24q=""
                  name="edit"
                  onClick={[Function]}
                  size="m"
                />
              </div>
            </div>
            <div
              data-css-b7lh05=""
            >
              <div
                data-css-1ajzdj4=""
                data-css-83ugar=""
              >
                <Icon
                  color="white"
                  data-css-1lke24q=""
                  name="remove-light-filled"
                  onClick={[Function]}
                  size="m"
                />
              </div>
            </div>
          </div>,
        ],
        "data-css-zh5k6q": "",
        "padded": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <_class
                checked={false}
                data-css-1mql7k=""
                id="checkmark0"
                onClick={[Function]}
              />,
              <div
                data-css-leaz66=""
                onDoubleClick={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                <Text
                  autoBreak={true}
                  block={true}
                  color="text"
                  data-css-93utwa=""
                  id={0}
                  size="l"
                >
                  Some Text
                </Text>
              </div>,
            ],
            "data-css-154b89o": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "checked": false,
                "data-css-1mql7k": "",
                "id": "checkmark0",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Text
                  autoBreak={true}
                  block={true}
                  color="text"
                  data-css-93utwa=""
                  id={0}
                  size="l"
                >
                  Some Text
                </Text>,
                "data-css-leaz66": "",
                "onDoubleClick": [Function],
                "onTouchMove": [Function],
                "onTouchStart": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "autoBreak": true,
                  "block": true,
                  "children": "Some Text",
                  "color": "text",
                  "data-css-93utwa": "",
                  "id": 0,
                  "size": "l",
                },
                "ref": null,
                "rendered": "Some Text",
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                data-css-b7lh05=""
              >
                <div
                  data-css-1ajzdj4=""
                  data-css-pmoomt=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="edit"
                    onClick={[Function]}
                    size="m"
                  />
                </div>
              </div>,
              <div
                data-css-b7lh05=""
              >
                <div
                  data-css-1ajzdj4=""
                  data-css-83ugar=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="remove-light-filled"
                    onClick={[Function]}
                    size="m"
                  />
                </div>
              </div>,
            ],
            "data-css-m0lcdg": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  data-css-1ajzdj4=""
                  data-css-pmoomt=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="edit"
                    onClick={[Function]}
                    size="m"
                  />
                </div>,
                "data-css-b7lh05": "",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="edit"
                    onClick={[Function]}
                    size="m"
                  />,
                  "data-css-1ajzdj4": "",
                  "data-css-pmoomt": "",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "color": "white",
                    "data-css-1lke24q": "",
                    "name": "edit",
                    "onClick": [Function],
                    "size": "m",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <div
                  data-css-1ajzdj4=""
                  data-css-83ugar=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="remove-light-filled"
                    onClick={[Function]}
                    size="m"
                  />
                </div>,
                "data-css-b7lh05": "",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="remove-light-filled"
                    onClick={[Function]}
                    size="m"
                  />,
                  "data-css-1ajzdj4": "",
                  "data-css-83ugar": "",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "color": "white",
                    "data-css-1lke24q": "",
                    "name": "remove-light-filled",
                    "onClick": [Function],
                    "size": "m",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                "type": "div",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ListItem
          backgroundColor="white"
          data-css-zh5k6q=""
          padded={true}
        >
          <div
            data-css-154b89o=""
          >
            <_class
              checked={false}
              data-css-1mql7k=""
              id="checkmark0"
              onClick={[Function]}
            />
            <div
              data-css-leaz66=""
              onDoubleClick={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
            >
              <Text
                autoBreak={true}
                block={true}
                color="text"
                data-css-93utwa=""
                id={0}
                size="l"
              >
                Some Text
              </Text>
            </div>
          </div>
          <div
            data-css-m0lcdg=""
          >
            <div
              data-css-b7lh05=""
            >
              <div
                data-css-1ajzdj4=""
                data-css-pmoomt=""
              >
                <Icon
                  color="white"
                  data-css-1lke24q=""
                  name="edit"
                  onClick={[Function]}
                  size="m"
                />
              </div>
            </div>
            <div
              data-css-b7lh05=""
            >
              <div
                data-css-1ajzdj4=""
                data-css-83ugar=""
              >
                <Icon
                  color="white"
                  data-css-1lke24q=""
                  name="remove-light-filled"
                  onClick={[Function]}
                  size="m"
                />
              </div>
            </div>
          </div>
        </ListItem>,
        "data-css-h0biv2": "",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "backgroundColor": "white",
          "children": Array [
            <div
              data-css-154b89o=""
            >
              <_class
                checked={false}
                data-css-1mql7k=""
                id="checkmark0"
                onClick={[Function]}
              />
              <div
                data-css-leaz66=""
                onDoubleClick={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
              >
                <Text
                  autoBreak={true}
                  block={true}
                  color="text"
                  data-css-93utwa=""
                  id={0}
                  size="l"
                >
                  Some Text
                </Text>
              </div>
            </div>,
            <div
              data-css-m0lcdg=""
            >
              <div
                data-css-b7lh05=""
              >
                <div
                  data-css-1ajzdj4=""
                  data-css-pmoomt=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="edit"
                    onClick={[Function]}
                    size="m"
                  />
                </div>
              </div>
              <div
                data-css-b7lh05=""
              >
                <div
                  data-css-1ajzdj4=""
                  data-css-83ugar=""
                >
                  <Icon
                    color="white"
                    data-css-1lke24q=""
                    name="remove-light-filled"
                    onClick={[Function]}
                    size="m"
                  />
                </div>
              </div>
            </div>,
          ],
          "data-css-zh5k6q": "",
          "padded": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <_class
                  checked={false}
                  data-css-1mql7k=""
                  id="checkmark0"
                  onClick={[Function]}
                />,
                <div
                  data-css-leaz66=""
                  onDoubleClick={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                >
                  <Text
                    autoBreak={true}
                    block={true}
                    color="text"
                    data-css-93utwa=""
                    id={0}
                    size="l"
                  >
                    Some Text
                  </Text>
                </div>,
              ],
              "data-css-154b89o": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "checked": false,
                  "data-css-1mql7k": "",
                  "id": "checkmark0",
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <Text
                    autoBreak={true}
                    block={true}
                    color="text"
                    data-css-93utwa=""
                    id={0}
                    size="l"
                  >
                    Some Text
                  </Text>,
                  "data-css-leaz66": "",
                  "onDoubleClick": [Function],
                  "onTouchMove": [Function],
                  "onTouchStart": [Function],
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "autoBreak": true,
                    "block": true,
                    "children": "Some Text",
                    "color": "text",
                    "data-css-93utwa": "",
                    "id": 0,
                    "size": "l",
                  },
                  "ref": null,
                  "rendered": "Some Text",
                  "type": [Function],
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <div
                  data-css-b7lh05=""
                >
                  <div
                    data-css-1ajzdj4=""
                    data-css-pmoomt=""
                  >
                    <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="edit"
                      onClick={[Function]}
                      size="m"
                    />
                  </div>
                </div>,
                <div
                  data-css-b7lh05=""
                >
                  <div
                    data-css-1ajzdj4=""
                    data-css-83ugar=""
                  >
                    <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="remove-light-filled"
                      onClick={[Function]}
                      size="m"
                    />
                  </div>
                </div>,
              ],
              "data-css-m0lcdg": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    data-css-1ajzdj4=""
                    data-css-pmoomt=""
                  >
                    <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="edit"
                      onClick={[Function]}
                      size="m"
                    />
                  </div>,
                  "data-css-b7lh05": "",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="edit"
                      onClick={[Function]}
                      size="m"
                    />,
                    "data-css-1ajzdj4": "",
                    "data-css-pmoomt": "",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "color": "white",
                      "data-css-1lke24q": "",
                      "name": "edit",
                      "onClick": [Function],
                      "size": "m",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "div",
                },
                "type": "div",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <div
                    data-css-1ajzdj4=""
                    data-css-83ugar=""
                  >
                    <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="remove-light-filled"
                      onClick={[Function]}
                      size="m"
                    />
                  </div>,
                  "data-css-b7lh05": "",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": <Icon
                      color="white"
                      data-css-1lke24q=""
                      name="remove-light-filled"
                      onClick={[Function]}
                      size="m"
                    />,
                    "data-css-1ajzdj4": "",
                    "data-css-83ugar": "",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "color": "white",
                      "data-css-1lke24q": "",
                      "name": "remove-light-filled",
                      "onClick": [Function],
                      "size": "m",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  "type": "div",
                },
                "type": "div",
              },
            ],
            "type": "div",
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
